rasa_core_sdk/forms.py

```diff
@@ -5,176 +5,197 @@
+                    intent = tracker.latest_message.get("intent",
+                                                        {}).get("name")
+                    if intent in slot_mapping.keys():
+                        return [SlotSet(slot_to_fill, slot_mapping[intent])]
+                else:
+                    entity_value = next(tracker.get_latest_entity_values(
+                                            slot_mapping), None)
+                    if entity_value is not None:
+                        return [SlotSet(slot_to_fill, entity_value)]
+
+            # the whole text can be always extracted, so it is done in the end
+            if self.FREETEXT in slot_mappings:
+                return [SlotSet(slot_to_fill,
+                                tracker.latest_message.get("text"))]
+
+        return None
+
+    # noinspection PyUnusedLocal
+    def validate(self, dispatcher, tracker, domain):
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """"Extract the user input else return an error"""
+        slot_to_fill = tracker.get_slot(REQUESTED_SLOT)
+
+        events = self.extract(dispatcher, tracker, domain)
+
+        if events is not None:
+            return events
         else:
-            return []
-
+            raise ActionExecutionRejected(self.name(),
+                                          "Failed to validate slot {0} "
+                                          "with action {1}"
+                                          "".format(slot_to_fill, self.name()))
+
+    # noinspection PyUnusedLocal
+    def request_next_slot(self,
+                             dispatcher,  # type: CollectingDispatcher
+                             tracker,  # type: Tracker
+                             domain  # type: Dict[Text, Any]
+                             ):
+        # type: (...) -> Optional[List[Dict]]
+        """"Request the next slot and utter template if needed,
+            else return None"""
+
+        for slot in self.required_slots():
+            if self._should_request_slot(tracker, slot):
+                dispatcher.utter_template("utter_ask_{}".format(slot), tracker)
+                return [SlotSet(REQUESTED_SLOT, slot)]
+
+        return None
 
-class BooleanFormField(FormField):
-    """A form field that prompts the user for a yes or no answer.
-    The interpreter should map positive and negative answers to
-    the intents ``affirm_intent`` and ``deny_intent``.
-    """
+    def submit(self, dispatcher, tracker, domain):
+        # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+        """Define what the form has to do
+            after all required slots are filled"""
 
-    def __init__(self, slot_name, affirm_intent, deny_intent):
-        self.slot_name = slot_name
-        self.affirm_intent = affirm_intent
-        self.deny_intent = deny_intent
+        raise NotImplementedError("A form must implement a submit method")
```

Comment:
@amn41 why `submit` raises `NotImplementedError`, why don't we use `pass` as default?

Reply:
I think in almost all cases you want to do something with the information once you collect it, this reminds the user that this `submit` will be called. If they really don't need it they can override it with a `pass`